### PR TITLE
Feat: Add defaults and verify exclusion handling.

### PR DESCRIPTION
**Pull request by steveworley**

### CODE REVIEW

File: modules/quant_purger/src/Plugin/Purge/Queuer/QuantPurger.php [hunk 1]
```diff
@@ -82,15 +82,9 @@ public function invalidateTags(array $tags) {
...
     }
-
-    $config = \Drupal::config('quant_purger.settings');
-
-    $blocklist = is_array($config->get('tag_blocklist')) ? $config->get('tag_blocklist') : [];
-
+    
     // Clean up the tags that we care about for this run.
     $tags = array_diff($tags, $this->invalidatedTags);
-    $tags = array_diff($tags, $blocklist);
```
Comment (steveworley) [inline modules/quant_purger/src/Plugin/Purge/Queuer/QuantPurger.php:92]: Moved this to the traffic registry, if tags or paths are matched the registry will not be updated with the matched request and won't be included as a candidate here.

File: modules/quant_purger/src/StackMiddleware/UrlRegistrar.php [hunk 2]
```diff
@@ -2,112 +2,127 @@
...
-   * The HTTP kernel object.
-   *
-   * @var \Symfony\Component\HttpKernel\HttpKernelInterface
-   */
-  protected $httpKernel;
+// In Drupal10+ HttpKernelInterface has been updated with
+// Symfony to have a different signature.
+//
+// @see https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/HttpKernel/HttpKernelInterface.php.
+if (empty($method->getReturnType())) {
```
Comment (steveworley) [inline modules/quant_purger/src/StackMiddleware/UrlRegistrar.php:18]: Now with more reflection — d9+d10 support in the same place.

Comment (kepol) [reply modules/quant_purger/src/StackMiddleware/UrlRegistrar.php:18]: I diffed the two classes:

```
41c41
<     public function handle(Request $request, $type = self::MASTER_REQUEST, $catch = TRUE) {
---
>     public function handle(Request $request, int $type = self::MAIN_REQUEST, bool $catch = TRUE): Response {
42a43
>       var_dump('d10');
```

I see this in 4.4 and lower:

https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/HttpKernel/HttpKernelInterface.php

```
    public function handle(Request $request, $type = self::MASTER_REQUEST, $catch = true);
```

I see this in 5.0+

```
    public function handle(Request $request, int $type = self::MAIN_REQUEST, bool $catch = true): Response;
```

so it looks like this matches the signatures... 

Need to remove the var_dump :)

Comment (kepol) [reply modules/quant_purger/src/StackMiddleware/UrlRegistrar.php:18]: I see the response was added in 9.3:

https://www.drupal.org/project/drupal/issues/3232082

not sure about the rest.

**Review by kepol — CHANGES_REQUESTED** [review]

**Comment by kepol** [comment]
Note that I didn't review anything else and didn't test anything. I'll scan the rest now.

File: modules/quant_purger/quant_purger.install [hunk 3]
```diff
@@ -34,3 +34,23 @@ function quant_purger_schema() {
   ];
   return $schema;
 }
+
+/**
+ * Update configuration for module settings.
+ */
+function quant_purger_update_0001(&$sandbox) {
```
Comment (kepol) [inline modules/quant_purger/quant_purger.install]: I'm used to update hooks being like `1000`, `1001`, etc. Does having it use `0001` do something special?

File: modules/quant_purger/quant_purger.install [hunk 4]
```diff
@@ -34,3 +34,23 @@ function quant_purger_schema() {
...
+  $config->set('tag_blocklist', [
+    'rendered',
+    'http_response',
+    '4xx-response',
+    'block_view',
+    'block_view_content',
+    'config:*',
+    'node_view',
+    'node_list',
+    'media_view',
```
Comment (kepol) [inline modules/quant_purger/quant_purger.install:81]: This doesn't completely match the settings file but maybe I'm unclear what's going on here.

**Review by kepol — COMMENTED** [review]
Added a couple more comments... I haven't tested yet.

File: modules/quant_purger/src/StackMiddleware/UrlRegistrar.php [hunk 5]
```diff
@@ -2,112 +2,127 @@
...
-      $this->registry->add(
-        $this->generateUrl($request),
-        $response->getCacheableMetadata()->getCacheTags()
-      );
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(Request $request, int $type = self::MAIN_REQUEST, bool $catch = TRUE): Response {
+      $response = $this->httpKernel->handle($request, $type, $catch);
+      var_dump('d10');
```
Comment (steveworley) [inline modules/quant_purger/src/StackMiddleware/UrlRegistrar.php]:
```suggestion
```

File: modules/quant_purger/src/StackMiddleware/UrlRegistrar.php (hunk 2 above)
Comment (steveworley) [reply modules/quant_purger/src/StackMiddleware/UrlRegistrar.php:18]: Fixed up the var_dump — should be good to go :D

**Comment by kepol** [comment]
Merging 8.x branch in to hopefully fix the mariadb issue.

**Comment by kepol** [comment]
That fixed it.

**Comment by kepol** [comment]
Update hook numbering to match documentation:
 https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_update_N/10

**Comment by kepol** [comment]
I must be testing this wrong because I can't get it to work. I tried:

1. Ran seed with nodes
2. Purge table was filled up with all the node URLs and cache tags

The particular page I was focused on was `/en/recipes/deep-mediterranean-quiche` from the Umami demo.

Test 1:

1. Clear queue
2. Add `media:1` to quant purge tag blocklist
4. Edited media 1 alt text
5. Process queue (through UI with nothing selected, maybe this is the problem)
6. No change to quant static page (expected)
7. Remove `media:1` to quant purge tag blocklist
8. Edited media 1 alt text
9. Process queue (through UI with nothing selected, maybe this is the problem)
10. No change to quant static page (not expected)

Test 2:

1. Clear queue
2. Edited taxonomy term `Main courses` (shows up in breadcrumb)
3. Process queue (through UI with nothing selected, maybe this is the problem)
4. No change to quant static page (not expected)

Test 3:

1. Clear queue
2. Edited footer block
3. Process queue (through UI with nothing selected, maybe this is the problem)
4. No change to quant static page (not expected)

I'll try again running queue via command line. It would be nice to be able to run the queue through the UI without anything chosen if there is stuff in the queue.

**Comment by kepol** [comment]
Yeah, that works. Okay... I'll make a feature request ticket for allowing processing the queue from the command line without checking any of the boxes... perhaps as a new button.

**Comment by kepol** [comment]
Spoke too soon. I added `block_content:*` to the tag blocklist but it still updated the content. Hmm... maybe I don't understand the point of the module.

**Comment by kepol** [comment]
Okay... I think there are a couple things going on here:

1. I don't think taxonomy terms are doing the right thing. I created a new taxonomy term that only one (new) node used. I added that term to the tag blocklist. I edited the term. It added all the nodes to the queue.

2. When I did a similar thing for media items and blocks then it worked when I had untranslated content, so I'm suspicious that if there are any translations then it still gets processed.

I've done a lot of fixes and testing and poking around so maybe I should stop at this point and find out how users are going to typically use this so I can focus on those use cases.

File: modules/quant_purger/quant_purger.install (hunk 3 above)
Comment (steveworley) [reply modules/quant_purger/quant_purger.install]: Nope, I just missed the convention when supporting multiple version of Drupal (it should use the lowest supported core version). As I understand them, the number doesn't really matter as long as its sequential, there are community guidelines on how to name them though.

> 1 or 2 digits for Drupal core compatibility (Drupal 8, 9, 10, etc.). This convention must be followed. If your module is compatible with multiple major versions (e.g., it has a core_version_requirement of '^8.8 || ^9'), use the lowest major core branch it is compatible with (8 in this example).

**Comment by steveworley** [comment]
Hey @kepol thanks for the fixes and the testing! I'll run through the test cases that I've been using to validate this. The idea behind the purger module is to:
- Track which URLs contain which tags and when Drupal initiates a cache tag invalidation it will collect all paths that have those tags and then add them to the queue for processing.

I've been running with the following assumptions:
- The main Quant module is installed and configured to seed content correctly
- The Quant purger module requires a full seed to populate the `purge_queuer_quant` table; which tracks the URL > Tag mapping

I've added one test case based on your testing and the issue raised:

For all test cases I run with the following setup tasks:
- Install Drupal (umami or else)
- Install and configure quant, quant API
-- Choose which entities Quant can operate on (for my tests I've been using nodes/taxonomies)
-- Configure Quant API to a test project in Quant

## `quant_purger` can be installed/uninstalled (added)

- `drush en quant_purger` — validate no errors
- `drush pmu quant_purger` — validate no errors

## Quant purger tracks URL and tags correctly

- Prepare a full seed `drush quant:seed-queue`
- Run the queue `drush quant:run-queue`

Query the database and validate that the tags are what I would expect (part of this change has been to include defaults that will reduce the surface area of the purger).

### Update the blocklist and validate

- Add `node:1` (or similar) to the tag blocklist
- truncate the `purge_queuer_quant` table
- Prepare/run the site seed

Validate that the tracking table reflects the change (eg. `select count(urlid) from purge_queuer_quant where tags like '%node:1%'` (should be 0)

## Validate that the queue is populated when tags are invalidated

- Prepare/run the site seed to populate the tracker table
- Find a tag that appears on many pages `select count(urlid) from purge_queuer_quant where tags like "%node:1%"` 

If the tag appears on more than one path, the next time `node:1` is invalidated it should include all impacted paths in the queue. For node 1 — edit node 1 to invalidate the tag and then check the queue.

These should all be passing now, at least from my expectation :) happy to discuss if you think the purger should work differently.

**Comment by kepol** [comment]
Thanks for the detailed notes. This is very helpful. I think this is where I got derailed:

1. Add `node:1` (or similar) to the tag blocklist
2. truncate the `purge_queuer_quant` table
3. Prepare/run the site seed

I did 1 and 3 but not 2. I was thinking that the `purge_queuer_quant` table had all the cache tags regardless of the blocklist and the blocklist was used when processing so that it would ignore processing when the tag was in the blocklist.

If I'm understanding correctly, the tag will be missing from the `purge_queuer_quant` table if the tag is added to the blocklist and then the db table is refreshed (truncate+seed).

This seems like it's easy to get wrong from a user perspective. Since I didn't realize I needed to do step 2, I imagine others would as well. I see 2 possible improvement options:

1. Restructure so that it's like my original assumption above. I haven't reviewed the code enough to know if this is a good idea as I assume it would require a lot of restructuring.

or

2. Automatically truncate the `purge_queuer_quant` table if any of the configuration changes rather than relying on the user to do this. And, add a warning message that explains that the `purge_queuer_quant` table is empty and requires a full seed. This is probably the easier approach.

I'll do some testing based on the steps above to make sure I do understand the process now.

**Comment by kepol** [comment]
I have confirmed that the install/uninstall error is fixed.

**Comment by kepol** [comment]
Tested adding node 1 tags in blocklist:
--

1. `node:1` - added to config then regenerated registry and edited node - node 1 not added to queue (expected) [note, turned off "Track content change" in order to edit the node and not trigger seed after edit]
1. `block_content:1` - added to config then regenerated registry and edited block - node 1 not added to queue (expected)
1. `taxonomy:31` - added to config then regenerated registry and edited term - node 1 not added to queue (expected)
1. `media:1` - added to config then regenerated registry and edited media - node 1 not added to queue (expected)
1. `file:37` - added to config then regenerated registry and edited (replaced) file - node 1 not added to queue (expected) [note, used file_management and file_management_view modules though not entirely clear if this would have invalidated the file:37 tag]

I am trying to test the opposite (that node 1 gets added to the queue when editing these things while these tags aren't in the blocklist) but I've run into an issue populating the `purge_queuer_quant` table that I'm debugging.

UPDATE: I couldn't figure out why it stopped working but I uninstalled and reinstalled and then it started working again.

**Comment by kepol** [comment]
Tested without node 1 tags in blocklist:
--

1. Removed tags tested above from blocklist and regenerated registry
1. `node:1` - edited node 1 and the right things were added to the queue
1. `block_content:1` - edited block content 1 and the right things were added to the queue
1. `taxonomy:31` - edited taxonomy term 31 and the right things were added to the queue
1. `media:1` - edited media item 1 and the right things were added to the queue
1. `file: 37` - edited file 37 and the right things were added to the queue

**Comment by kepol** [comment]
Ok... it seems to be working as you have described. Yay! But I did have some UX concerns noted in:

https://github.com/quantcdn/drupal/pull/155#issuecomment-1654898977

Note also that the cancel button isn't working: https://www.drupal.org/project/quantcdn/issues/3377627

I did fix the clear registry button above

File: modules/quant_purger/quant_purger.install (hunk 3 above)
Comment (kepol) [reply modules/quant_purger/quant_purger.install]: I followed something like this one: 

```
mymodule_update_8201(): The first update for the custom or contributed module's 2.0.x versions, which are compatible with Drupal 8 and 9.
```

from https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_update_N/9

File: modules/quant_purger/src/Plugin/Purge/Queuer/QuantPurger.php (hunk 1 above)
Comment (kepol) [reply modules/quant_purger/src/Plugin/Purge/Queuer/QuantPurger.php:92]: I have tested and things appear to be working properly so marking resolved.

**Comment by kepol** [comment]
Cool. I'll retest where I don't regenerate the table and make sure these also work.

**Comment by kepol** [comment]
Fixed error because `$config` not in class.

```
Error: Call to a member function get() on null in Drupal\quant_purger\TrafficRegistry->getAcceptedCacheTags() (line 80 of /var/www/html/web/modules/contrib/quantcdn/modules/quant_purger/src/StackMiddleware/TraitUrlRegistrar.php)
```

**Comment by kepol** [comment]
Fixes and UX improvements above ^. Also tested the new changes by:

1. Run full seed
1. `node:1` - added to config and did *not* regenerate registry and edited node - node 1 not added to queue (expected) [note, turned off "Track content change" in order to edit the node and not trigger seed after edit]
1. `block_content:1` - added to config and did *not* regenerate registry and edited block - node 1 not added to queue (expected)
1. `taxonomy:31` - added to config and did *not* regenerate registry and edited term - node 1 not added to queue (expected)
1. `media:19` - added to config and did *not* regenerate registry and edited media - node 1 not added to queue (expected)
1. `file:37` - added to config and did *not* regenerate registry and edited (replaced) file - node 1 not added to queue (expected) [note, used file_management and file_management_view modules though not entirely clear if this would have invalidated the `file:37` tag]

The only thing that didn't work above was for `taxonomy:31`. The taxonomy term page still showed up in the queue even though the cache tag was present but I assume this is because it's getting pushed upon edit due to this bug: https://www.drupal.org/project/quantcdn/issues/3342521

```
There are [2] items in the queue.

Item ID	Claimed/Expiration	Created	Content/Data
14683	unclaimed	Sun, 30 Jul 2023 23:27:41 +0000	Route: /en/recipe-category/main-courses
14684	unclaimed	Sun, 30 Jul 2023 23:27:41 +0000	Route: /es/recipe-category/platos-principales
```

```
|   282 | /en/recipe-category/main-courses                | ;block_content:2;block_content:1;taxonomy_term:31;media:19;file:37;taxonomy_term:16;taxonomy_term:15;taxonomy_term:14;taxonomy_term:13;taxonomy_term:12;taxonomy_term:11;taxonomy_term:10;taxonomy_term:9;taxonomy_term:8;taxonomy_term:7;taxonomy_term:6;taxonomy_term:5;taxonomy_term:4;taxonomy_term:3;taxonomy_term:2;taxonomy_term:1;node:1;node:3;node:6;node:7;node:10;media:18;media:3;media:6;media:7;media:21;taxonomy_term_view |
|   283 | /es/recipe-category/platos-principales          | ;block_content:2;block_content:1;taxonomy_term:31;media:19;file:38;taxonomy_term:16;taxonomy_term:15;taxonomy_term:14;taxonomy_term:13;taxonomy_term:12;taxonomy_term:11;taxonomy_term:10;taxonomy_term:9;taxonomy_term:8;taxonomy_term:7;taxonomy_term:6;taxonomy_term:5;taxonomy_term:4;taxonomy_term:3;taxonomy_term:2;taxonomy_term:1;node:1;node:3;node:6;node:7;node:10;media:3;media:6;media:7;media:21;taxonomy_term_view          |
```

**Comment by kepol** [comment]
@steveworley I think this is good enough for this round. Check the changes I made above.

UPDATE: I snuck in one more thing below.

**Comment by kepol** [comment]
Steve took a look and said things look good to him so I'll get this merged in and make a release soon.

**Review by kepol — APPROVED** [review]
This looks good and has been thoroughly tested. Any remaining issues/improvements found with purger functionality have issues in the drupal.org queue.

**Comment by kepol** [comment]
Released: https://github.com/quantcdn/drupal/releases/tag/1.5.2

**Pull request merged**